File: api/services/mood_service.py
from typing import List, Optional, Dict
from api.database import MoodDatabase
from api.models.mood_entry import MoodEntry


class MoodService:
    def __init__(self, db: MoodDatabase):
        self.db = db

    def create_mood_entry(
        self,
        user_id: int,
        date: str,
        mood: int,
        content: str,
        time: Optional[str] = None,
        selected_options: Optional[List[int]] = None,
        slider_values: Optional[Dict[int, int]] = None,
        is_important: Optional[bool] = None,
        important_reason: Optional[str] = None,
    ) -> Dict:
        """Create a new mood entry and check for achievements"""
        if not (1 <= mood <= 5):
            raise ValueError("Mood must be between 1 and 5")

        if not content.strip():
            raise ValueError("Content cannot be empty")

        if is_important and not (important_reason or "").strip():
            raise ValueError("Please provide a reason for marking this day important")

        entry_id = self.db.add_mood_entry(
            user_id, date, mood, content, time, selected_options, slider_values,
            is_important, important_reason,
        )

        # Check for new achievements
        new_achievements = self.db.check_achievements(user_id)

        return {"entry_id": entry_id, "new_achievements": new_achievements}

    def get_all_entries(self, user_id: int) -> List[Dict]:
        """Get all mood entries for a user"""
        return self.db.get_all_mood_entries(user_id)

    def get_entries_by_date_range(
        self, user_id: int, start_date: str, end_date: str
    ) -> List[Dict]:
        """Get mood entries within a date range for a user"""
        return self.db.get_mood_entries_by_date_range(user_id, start_date, end_date)

    def get_entry_by_id(self, user_id: int, entry_id: int) -> Optional[Dict]:
        """Get a specific mood entry by ID for a user"""
        return self.db.get_mood_entry_by_id(user_id, entry_id)

    def update_entry(
        self,
        user_id: int,
        entry_id: int,
        mood: Optional[int] = None,
        content: Optional[str] = None,
        date: Optional[str] = None,
        time: Optional[str] = None,
        selected_options: Optional[List[int]] = None,
        slider_values: Optional[Dict[int, int]] = None,
        is_important: Optional[bool] = None,
        important_reason: Optional[str] = None,
    ) -> Optional[Dict]:
        """Update an existing mood entry for a user and return the updated record"""
        if mood is not None and not (1 <= mood <= 5):
            raise ValueError("Mood must be between 1 and 5")

        if content is not None and not content.strip():
            raise ValueError("Content cannot be empty")

        if is_important and not (important_reason or "").strip():
            raise ValueError("Please provide a reason for marking this day important")

        updated = self.db.update_mood_entry(
            user_id,
            entry_id,
            mood=mood,
            content=content,
            date=date,
            time=time,
            selected_options=selected_options,
            slider_values=slider_values,
            is_important=is_important,
            important_reason=important_reason,
        )

        if not updated:
            return None

        entry = self.db.get_mood_entry_by_id(user_id, entry_id)
        if not entry:
            return None

        entry["selections"] = self.db.get_entry_selections(entry_id)
        entry["slider_values"] = self.db.get_entry_slider_values(entry_id)
        return entry

    def add_mood_log(self, user_id: int, entry_id: int, mood: int) -> Dict:
        """Add a mood check-in log to an entry and return the updated mood."""
        if not (1 <= mood <= 5):
            raise ValueError("Mood must be between 1 and 5")
        return self.db.add_mood_log(user_id, entry_id, mood)

    def get_mood_logs(self, user_id: int, entry_id: int) -> List[Dict]:
        """Return all mood logs for an entry, ordered chronologically."""
        return self.db.get_mood_logs(user_id, entry_id)

    def delete_mood_log(self, user_id: int, entry_id: int, log_id: int) -> Optional[Dict]:
        """Delete a mood log and return the updated entry mood, or None if not found."""
        new_mood = self.db.delete_mood_log(user_id, entry_id, log_id)
        if new_mood is None:
            return None
        return {"updated_entry_mood": new_mood}

    def delete_entry(self, user_id: int, entry_id: int) -> bool:
        """Delete a mood entry for a user"""
        return self.db.delete_mood_entry(user_id, entry_id)

    def get_statistics(self, user_id: int) -> Dict:
        """Get mood statistics for a user"""
        # Track statistics view for achievements (Data Lover)
        try:
            self.db.increment_stats_view(user_id)
        except Exception:
            # Metrics should not break stats
            pass
        stats = self.db.get_mood_statistics(user_id)
        mood_counts = self.db.get_mood_counts(user_id)
        current_streak = self.db.get_current_streak(user_id)

        return {
            "statistics": stats,
            "mood_distribution": mood_counts,
            "current_streak": current_streak,
        }

    def get_current_streak(self, user_id: int) -> int:
        """Get current consecutive days streak for a user"""
        return self.db.get_current_streak(user_id)

    def get_entry_selections(self, user_id: int, entry_id: int) -> List[Dict]:
        """Get selected options for an entry (with user verification)"""
        # First verify the entry belongs to the user
        entry = self.db.get_mood_entry_by_id(user_id, entry_id)
        if not entry:
            return []
        return self.db.get_entry_selections(entry_id)

    def get_entry_slider_values(self, user_id: int, entry_id: int) -> List[Dict]:
        """Get slider values for an entry (with user verification)"""
        entry = self.db.get_mood_entry_by_id(user_id, entry_id)
        if not entry:
            return []
        return self.db.get_entry_slider_values(entry_id)

File: functions/_shared/services/mood-service.ts
import { and, asc, between, desc, eq, inArray, sql } from 'drizzle-orm';
import type { BatchItem } from 'drizzle-orm/batch';
import { toBatch, type Database } from '../db/client';
import {
  entryMoodLogs,
  entrySelections,
  entrySliderValues,
  groupOptions,
  groups,
  moodEntries,
} from '../db/schema';
import type { MoodCreateInput, MoodUpdateInput } from '@shared/schemas/mood';
import {
  checkAchievements,
  getCurrentStreak,
  getMoodCounts,
  getMoodStatistics,
  incrementStatsView,
} from './achievement-service';

export class ValidationError extends Error {}

export interface MoodEntryRecord {
  id: number;
  date: string;
  mood: number;
  content: string;
  created_at: string;
  updated_at: string;
  is_important: boolean;
  important_reason: string | null;
}

export interface EntrySelectionRecord {
  id: number;
  group_id: number;
  name: string;
  icon: string | null;
  group_name: string;
  group_color: string | null;
}

export interface EntrySliderValueRecord {
  id: number;
  group_id: number;
  value: number;
  group_name: string;
  group_color: string | null;
  group_icon: string | null;
  slider_min: number | null;
  slider_max: number | null;
  slider_labels: string[] | null;
}

export interface MoodLogRecord {
  id: number;
  entry_id: number;
  mood: number;
  logged_at: string;
}

function toEntryRecord(row: typeof moodEntries.$inferSelect): MoodEntryRecord {
  return {
    id: row.id,
    date: row.date,
    mood: row.mood,
    content: row.content,
    created_at: row.createdAt,
    updated_at: row.updatedAt,
    is_important: row.isImportant,
    important_reason: row.importantReason,
  };
}

function assertImportantReason(isImportant: boolean | undefined, importantReason: string | undefined | null): void {
  if (isImportant && !(importantReason ?? '').trim()) {
    throw new ValidationError('Please provide a reason for marking this day important');
  }
}

async function assertUserOwnedEntryFields(
  db: Database,
  userId: number,
  selectedOptions: number[],
  sliderValues: Record<number, number>
): Promise<void> {
  const uniqueOptionIds = [...new Set(selectedOptions)];
  if (uniqueOptionIds.length > 0) {
    const rows = await db
      .select({ id: groupOptions.id })
      .from(groupOptions)
      .innerJoin(groups, eq(groupOptions.groupId, groups.id))
      .where(and(eq(groups.userId, userId), inArray(groupOptions.id, uniqueOptionIds)));
    if (rows.length !== uniqueOptionIds.length) {
      throw new ValidationError('One or more selected options are not available for this user');
    }
  }

  const groupIds = [...new Set(Object.keys(sliderValues).map(Number).filter(Number.isFinite))];
  if (groupIds.length > 0) {
    const rows = await db
      .select({ id: groups.id })
      .from(groups)
      .where(and(eq(groups.userId, userId), inArray(groups.id, groupIds)));
    if (rows.length !== groupIds.length) {
      throw new ValidationError('One or more sliders are not available for this user');
    }
  }
}

export async function getEntrySelections(db: Database, entryId: number): Promise<EntrySelectionRecord[]> {
  const rows = await db
    .select({
      id: groupOptions.id,
      group_id: groupOptions.groupId,
      name: groupOptions.name,
      icon: groupOptions.icon,
      group_name: groups.name,
      group_color: groups.color,
      group_sort_order: groups.sortOrder,
      option_sort_order: groupOptions.sortOrder,
    })
    .from(entrySelections)
    .innerJoin(groupOptions, eq(entrySelections.optionId, groupOptions.id))
    .innerJoin(groups, eq(groupOptions.groupId, groups.id))
    .where(eq(entrySelections.entryId, entryId))
    .orderBy(asc(groups.sortOrder), asc(groups.name), asc(groupOptions.sortOrder), asc(groupOptions.name));

  return rows.map(({ group_sort_order: _group_sort_order, option_sort_order: _option_sort_order, ...rest }) => rest);
}

export async function getEntrySliderValues(db: Database, entryId: number): Promise<EntrySliderValueRecord[]> {
  const rows = await db
    .select({
      id: entrySliderValues.id,
      group_id: entrySliderValues.groupId,
      value: entrySliderValues.value,
      group_name: groups.name,
      group_color: groups.color,
      group_icon: groups.icon,
      slider_min: groups.sliderMin,
      slider_max: groups.sliderMax,
      slider_labels: groups.sliderLabels,
      group_sort_order: groups.sortOrder,
    })
    .from(entrySliderValues)
    .innerJoin(groups, eq(entrySliderValues.groupId, groups.id))
    .where(eq(entrySliderValues.entryId, entryId))
    .orderBy(asc(groups.sortOrder), asc(groups.name));

  return rows.map(({ group_sort_order: _group_sort_order, ...rest }) => rest);
}

export async function createMoodEntry(
  db: Database,
  userId: number,
  input: MoodCreateInput,
  opts?: { skipAchievements?: boolean }
) {
  assertImportantReason(input.is_important, input.important_reason);

  const isImportant = Boolean(input.is_important);
  const selectedOptions = input.selected_options ?? [];
  const sliderValues = input.slider_values ?? {};
  await assertUserOwnedEntryFields(db, userId, selectedOptions, sliderValues);

  const [inserted] = await db
    .insert(moodEntries)
    .values({
      userId,
      date: input.date,
      mood: input.mood,
      content: input.content,
      ...(input.time ? { createdAt: input.time } : {}),
      isImportant,
      importantReason: input.important_reason ?? null,
    })
    .returning({ id: moodEntries.id });

  const entryId = inserted.id;

  // D1 has no cross-statement transactions reachable from Drizzle (see toBatch);
  // the dependent rows reference `entryId`, so they're inserted in one atomic
  // batch after the parent row exists, with a best-effort cleanup on failure.
  const dependentInserts: BatchItem<'sqlite'>[] = [];
  if (selectedOptions.length > 0) {
    dependentInserts.push(
      db.insert(entrySelections).values(selectedOptions.map((optionId) => ({ entryId, optionId })))
    );
  }

  const sliderEntries = Object.entries(sliderValues);
  if (sliderEntries.length > 0) {
    dependentInserts.push(
      db.insert(entrySliderValues).values(
        sliderEntries.map(([groupId, value]) => ({
          entryId,
          groupId: Number(groupId),
          value,
        }))
      )
    );
  }

  dependentInserts.push(
    db.insert(entryMoodLogs).values({
      entryId,
      mood: input.mood,
      ...(input.time ? { loggedAt: input.time } : {}),
    })
  );

  try {
    await db.batch(toBatch(dependentInserts));
  } catch (err) {
    await db.delete(moodEntries).where(eq(moodEntries.id, entryId));
    throw err;
  }

  const newAchievements = opts?.skipAchievements ? [] : await checkAchievements(db, userId);

  return { entry_id: entryId, new_achievements: newAchievements };
}

export async function getAllEntries(db: Database, userId: number): Promise<MoodEntryRecord[]> {
  const rows = await db
    .select()
    .from(moodEntries)
    .where(eq(moodEntries.userId, userId))
    .orderBy(desc(moodEntries.createdAt), desc(moodEntries.date));
  return rows.map(toEntryRecord);
}

export async function getEntriesByDateRange(
  db: Database,
  userId: number,
  startDate: string,
  endDate: string
): Promise<MoodEntryRecord[]> {
  const rows = await db
    .select()
    .from(moodEntries)
    .where(and(eq(moodEntries.userId, userId), between(moodEntries.date, startDate, endDate)))
    .orderBy(desc(moodEntries.createdAt), desc(moodEntries.date));
  return rows.map(toEntryRecord);
}

export async function getEntryById(db: Database, userId: number, entryId: number): Promise<MoodEntryRecord | null> {
  const row = await db.query.moodEntries.findFirst({
    where: and(eq(moodEntries.id, entryId), eq(moodEntries.userId, userId)),
  });
  return row ? toEntryRecord(row) : null;
}

export async function updateEntry(db: Database, userId: number, entryId: number, input: MoodUpdateInput) {
  assertImportantReason(input.is_important, input.important_reason);

  const existing = await db.query.moodEntries.findFirst({
    where: and(eq(moodEntries.id, entryId), eq(moodEntries.userId, userId)),
    columns: { id: true },
  });
  if (!existing) {
    return null;
  }

  if (input.selected_options !== undefined || input.slider_values !== undefined) {
    await assertUserOwnedEntryFields(db, userId, input.selected_options ?? [], input.slider_values ?? {});
  }

  const updates: Record<string, unknown> = {};
  if (input.mood !== undefined) updates.mood = input.mood;
  if (input.content !== undefined) updates.content = input.content;
  if (input.date !== undefined) updates.date = input.date;
  if (input.time !== undefined) updates.createdAt = input.time;
  if (input.is_important !== undefined) updates.isImportant = input.is_important;
  if (input.important_reason !== undefined) updates.importantReason = input.important_reason;
  updates.updatedAt = sql`CURRENT_TIMESTAMP`;

  const statements: BatchItem<'sqlite'>[] = [
    db
      .update(moodEntries)
      .set(updates)
      .where(and(eq(moodEntries.id, entryId), eq(moodEntries.userId, userId))),
  ];

  if (input.selected_options !== undefined) {
    statements.push(db.delete(entrySelections).where(eq(entrySelections.entryId, entryId)));
    if (input.selected_options.length > 0) {
      statements.push(
        db.insert(entrySelections).values(
          input.selected_options.map((optionId) => ({ entryId, optionId }))
        )
      );
    }
  }

  if (input.slider_values !== undefined) {
    statements.push(db.delete(entrySliderValues).where(eq(entrySliderValues.entryId, entryId)));
    const sliderEntries = Object.entries(input.slider_values);
    if (sliderEntries.length > 0) {
      statements.push(
        db.insert(entrySliderValues).values(
          sliderEntries.map(([groupId, value]) => ({
            entryId,
            groupId: Number(groupId),
            value,
          }))
        )
      );
    }
  }

  await db.batch(toBatch(statements));

  const entry = await getEntryById(db, userId, entryId);
  if (!entry) return null;

  const selections = await getEntrySelections(db, entryId);
  const sliderValues = await getEntrySliderValues(db, entryId);

  return { ...entry, selections, slider_values: sliderValues };
}

export async function deleteEntry(db: Database, userId: number, entryId: number): Promise<boolean> {
  const result = await db
    .delete(moodEntries)
    .where(and(eq(moodEntries.id, entryId), eq(moodEntries.userId, userId)))
    .returning({ id: moodEntries.id });
  return result.length > 0;
}

async function recalculateAverageMood(db: Database, entryId: number): Promise<number | null> {
  const [row] = await db
    .select({ avgMood: sql<number | null>`ROUND(AVG(${entryMoodLogs.mood}))` })
    .from(entryMoodLogs)
    .where(eq(entryMoodLogs.entryId, entryId));

  if (!row || row.avgMood == null) return null;

  const mood = Math.trunc(row.avgMood);
  await db
    .update(moodEntries)
    .set({ mood, updatedAt: sql`CURRENT_TIMESTAMP` })
    .where(eq(moodEntries.id, entryId));

  return mood;
}

export async function addMoodLog(db: Database, userId: number, entryId: number, mood: number) {
  const entry = await db.query.moodEntries.findFirst({
    where: and(eq(moodEntries.id, entryId), eq(moodEntries.userId, userId)),
    columns: { id: true },
  });
  if (!entry) {
    throw new ValidationError('Entry not found');
  }

  // Sequential read-then-write (recalc depends on the row just inserted, and the
  // response depends on the row recalc just updated) — D1 has no cross-statement
  // transaction primitive reachable from Drizzle that allows intermediate reads
  // (see toBatch), so these run as separate auto-committed statements.
  const [inserted] = await db
    .insert(entryMoodLogs)
    .values({ entryId, mood })
    .returning({ id: entryMoodLogs.id, mood: entryMoodLogs.mood, loggedAt: entryMoodLogs.loggedAt });

  await recalculateAverageMood(db, entryId);

  const [updatedEntry] = await db
    .select({ mood: moodEntries.mood })
    .from(moodEntries)
    .where(eq(moodEntries.id, entryId));

  return {
    log_id: inserted.id,
    mood: inserted.mood,
    logged_at: inserted.loggedAt,
    updated_entry_mood: updatedEntry?.mood ?? mood,
  };
}

export async function getMoodLogs(db: Database, userId: number, entryId: number): Promise<MoodLogRecord[]> {
  const entry = await db.query.moodEntries.findFirst({
    where: and(eq(moodEntries.id, entryId), eq(moodEntries.userId, userId)),
    columns: { id: true },
  });
  if (!entry) {
    throw new ValidationError('Entry not found');
  }

  const rows = await db
    .select({
      id: entryMoodLogs.id,
      entry_id: entryMoodLogs.entryId,
      mood: entryMoodLogs.mood,
      logged_at: entryMoodLogs.loggedAt,
    })
    .from(entryMoodLogs)
    .where(eq(entryMoodLogs.entryId, entryId))
    .orderBy(asc(entryMoodLogs.loggedAt));

  return rows;
}

export async function deleteMoodLog(
  db: Database,
  userId: number,
  entryId: number,
  logId: number
): Promise<number | null> {
  const entry = await db.query.moodEntries.findFirst({
    where: and(eq(moodEntries.id, entryId), eq(moodEntries.userId, userId)),
    columns: { id: true },
  });
  if (!entry) {
    return null;
  }

  const deleted = await db
    .delete(entryMoodLogs)
    .where(and(eq(entryMoodLogs.id, logId), eq(entryMoodLogs.entryId, entryId)))
    .returning({ id: entryMoodLogs.id });

  if (deleted.length === 0) {
    return null;
  }

  const recalculatedMood = await recalculateAverageMood(db, entryId);
  if (recalculatedMood !== null) return recalculatedMood;

  const [updatedEntry] = await db
    .select({ mood: moodEntries.mood })
    .from(moodEntries)
    .where(eq(moodEntries.id, entryId));

  return updatedEntry?.mood ?? null;
}

export async function getStatistics(db: Database, userId: number) {
  // Track statistics view for achievements (Data Lover); metrics must not break stats.
  try {
    await incrementStatsView(db, userId);
  } catch {
    // ignore
  }

  const [statistics, moodDistribution, currentStreak] = await Promise.all([
    getMoodStatistics(db, userId),
    getMoodCounts(db, userId),
    getCurrentStreak(db, userId),
  ]);

  return {
    statistics,
    mood_distribution: moodDistribution,
    current_streak: currentStreak,
  };
}

export { getCurrentStreak };
